### PR TITLE
EOS-15665 CSM User Password being fetched from Provisioner.

### PR DESCRIPTION
**Pull request by prathameshrodi**
# Backend

## Problem Statement
<pre>
Story Ref (if any): EOS-15665
Security: disable SSH or ensure strong passwords for all service accounts
</pre>
## Unit testing on RPM done
<pre>
No
</pre>
## Problem Description
<pre>
CSM Hardcoded Password are visible in Opensource Hence we will be fetching a randomly generating password.
</pre>
## Solution
<pre>
Changes made for fetching code from provisioner and saving encrypted randomly generated password.
</pre>
## Unit Test Cases
<pre>

</pre>
Signed-off-by:

### CODE REVIEW

**Review by kupranay — APPROVED** [review]

File: csm/core/blogic/const.py [hunk 1]
```diff
@@ -490,8 +490,12 @@
 AVAILABLE = 'avail'
 USAGE_PERCENTAGE = 'usage_percentage'
 
+# CSM conf entry
+CSM='CSM'
+
 # Keys for  Description
 DECRYPTION_KEYS = {
+    "CSM.password": "csm",
```
Comment (ajaysrivas) [inline csm/core/blogic/const.py:498]: NON_ROOT_USER_PASS is still there const.py.

File: csm/conf/setup.py [hunk 2]
```diff
@@ -160,8 +185,13 @@ def _config_user(self, reset=False):
         """
         if not reset:
             if not self._is_user_exist():
-                Setup._run_cmd("useradd -d "+const.CSM_USER_HOME+" -p "+self._password+" "+ self._user)
+                _password = self._fetch_csm_user_password(decrypt=True)
```
Comment (ajaysrivas) [inline csm/conf/setup.py:188]: Will these code changes work in DEV environment?

File: csm/conf/setup.py [hunk 3]
```diff
@@ -160,8 +185,13 @@ def _config_user(self, reset=False):
...
+                if not _password:
+                    Log.error("CSM Password Not Recieved from provisioner.")
+                    raise CsmSetupError("CSM Password Not Set by Provisioner.")
+                Setup._run_cmd("useradd -d "+const.CSM_USER_HOME+" -p "+_password+" "+ self._user)
                 Setup._run_cmd("usermod -aG wheel " + self._user)
+                Setup._run_cmd("usermod -s /sbin/nologin " + self._user)
```
Comment (navalkp) [inline csm/conf/setup.py:194]: You can remove this, this was part of alternate solution.

Comment (sachinpunadikar) [reply csm/conf/setup.py:194]: "nologin" is required. Do not remove it.

File: csm/conf/setup.py [hunk 4]
```diff
@@ -247,14 +279,21 @@ def store_encrypted_password(conf_data):
...
+            _paswd = Setup._fetch_csm_user_password()
+            if not _paswd:
+                raise CsmSetupError("CSM Password Not Set by Provisioner.")
+            Log.info("CSM Credentials Copied to CSM Configuration.")
+            conf_data[const.CSM][const.PASSWORD] = _paswd
             cluster_id = SaltWrappers.get_salt_call(const.GRAINS_GET, const.CLUSTER_ID)
             provisioner_data = conf_data[const.PROVISIONER]
             provisioner_data[const.CLUSTER_ID] = cluster_id
             conf_data[const.PROVISIONER] = provisioner_data
+            # NKP get csm user/password using salt command and set to conf
```
Comment (pawankumarsrivastava) [inline csm/conf/setup.py:296]: @navalkp Please check whether this comment is required. If not remove it.

File: csm/conf/setup.py [hunk 5]
```diff
@@ -716,6 +755,7 @@ def post_install(self, args):
         """
         try:
             self._verify_args(args)
+            # NKP Get csm user credential from salt and decrypt password and set to variables
```
Comment (pawankumarsrivastava) [inline csm/conf/setup.py:758]: @navalkp Please check whether this comment is required. If not remove it.

File: csm/conf/setup.py [hunk 6]
```diff
@@ -78,6 +78,31 @@ def _run_cmd(cmd):
...
             raise CsmSetupError("Csm setup is failed Error: %s %s" %(e,_err))
 
+    @staticmethod
+    def _fetch_csm_user_password(decrypt=False):
+        """
+        This Method Fetches the Password for CSM User from Provisioner.
+        :param decrypt:
+        :return:
+        """
+        Log.info("Fetch CSM User Password from provisioner.")
```
Comment (pawankumarsrivastava) [inline csm/conf/setup.py:88]: Fetch -> Fetching

File: csm/conf/setup.py [hunk 7]
```diff
@@ -78,6 +78,31 @@ def _run_cmd(cmd):
...
+        csm_user_pass = None
+        csm_credentials = SaltWrappers.get_salt_call(const.PILLAR_GET, "csm")
+        if csm_credentials and type(csm_credentials) is dict:
+            csm_user_pass = csm_credentials.get(const.SECRET)
+        else:
+            Log.error("No Credentials Fetched from Provisioner.")
```
Comment (pawankumarsrivastava) [inline csm/conf/setup.py:94]: If no credentials are fetched we should return from here. In this case we will save one if statement check.

File: csm/conf/setup.py [hunk 8]
```diff
@@ -78,6 +78,31 @@ def _run_cmd(cmd):
...
+        if decrypt and csm_user_pass:
+            Log.info("Decrypting CSM Password.")
+            cluster_id = SaltWrappers.get_salt_call(const.GRAINS_GET, const.CLUSTER_ID)
```
Comment (pawankumarsrivastava) [inline csm/conf/setup.py:97]: Any salt calls must be kept inside try except block in case it fails.

File: csm/conf/setup.py [hunk 9]
```diff
@@ -78,6 +78,31 @@ def _run_cmd(cmd):
...
+    @staticmethod
+    def _fetch_csm_user_password(decrypt=False):
+        """
+        This Method Fetches the Password for CSM User from Provisioner.
+        :param decrypt:
+        :return:
+        """
+        Log.info("Fetch CSM User Password from provisioner.")
+        csm_user_pass = None
+        csm_credentials = SaltWrappers.get_salt_call(const.PILLAR_GET, "csm")
```
Comment (pawankumarsrivastava) [inline csm/conf/setup.py:90]: Any salt calls must be kept inside try except block in case it fails.

File: csm/conf/setup.py [hunk 10]
```diff
@@ -236,9 +266,11 @@ class Config:
         @staticmethod
         def store_encrypted_password(conf_data):
             # read username's and password's for S3 and RMQ
+            Log.info("Store Encrypted Password")
```
Comment (pawankumarsrivastava) [inline csm/conf/setup.py:269]: Store -> Storing

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/blogic/const.py (hunk 1 above)
Comment (prathameshrodi) [reply csm/core/blogic/const.py:498]: yes will remove it.

File: csm/conf/setup.py (hunk 2 above)
Comment (prathameshrodi) [reply csm/conf/setup.py:188]: No these are not expected.

File: csm/conf/setup.py (hunk 4 above)
Comment (prathameshrodi) [reply csm/conf/setup.py:296]: will remove.

File: csm/conf/setup.py (hunk 5 above)
Comment (prathameshrodi) [reply csm/conf/setup.py:758]: will remove

File: csm/conf/setup.py (hunk 6 above)
Comment (prathameshrodi) [reply csm/conf/setup.py:88]: Will Change

File: csm/conf/setup.py (hunk 8 above)
Comment (prathameshrodi) [reply csm/conf/setup.py:97]: sure

File: csm/conf/setup.py (hunk 10 above)
Comment (prathameshrodi) [reply csm/conf/setup.py:269]: will change

Comment (prathameshrodi) [reply csm/conf/setup.py:269]: will change

File: csm/conf/setup.py (hunk 7 above)
Comment (prathameshrodi) [reply csm/conf/setup.py:94]: sure

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by navalkp — APPROVED** [review]

File: csm/conf/setup.py [hunk 11]
```diff
@@ -201,8 +201,11 @@ def _config_user(self, reset=False):
                 if not _password:
                     Log.error("CSM Password Not Recieved from provisioner.")
                     raise CsmSetupError("CSM Password Not Set by Provisioner.")
+                Log.info("Creating CSM User.")
                 Setup._run_cmd(f"useradd -d {const.CSM_USER_HOME} -p {_password} {self._user}")
+                Log.info("Adding CSM User to Wheel Group.")
                 Setup._run_cmd("usermod -aG wheel " + self._user)
+                Log.info("Adding CSM User to nologin Group.")
```
Comment (sachinpunadikar) [inline csm/conf/setup.py:208]: Change the comment to:
Enabling nologin for CSM user.

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
A minor change in the log statement required

**Comment by prathameshrodi** [comment]
@sachinpunadikar  Changed Log Statement

**Review by sachinpunadikar — APPROVED** [review]
LGTM

**Pull request merged**